docs/design_proposals/ko-builder.md

```diff
@@ -0,0 +1,240 @@
+  ko in Skaffold, ko will be a tool that all developers in a team would have to
+  install.
+
+- works great as a _CI/CD building block_: when using the ko builder, pipeline
+  steps can run using the default Skaffold container image, without
+  installing additional tools or keeping toolchain versions in sync across
+  local development and CI/CD.
+
+## Design
+
+Adding the ko builder requires making config changes to the Skaffold schema.
+
+1.  Add an entry to BuilderType enum in `proto/enums/enums.proto`:
+
+    ```proto
+        // ko Builder
+        KO = 7;
+    ```
+
+2.  Add a `KoArtifact` type:
+
+    ```go
+    type KoArtifact struct {
+    	// Annotations are key-value string pairs to add to the image manifest.
+    	// Also known as `LABEL` in `Dockerfile`s.
+    	// Ref: https://github.com/opencontainers/image-spec/blob/master/annotations.md
+    	Annotations map[string]string `yaml:"annotations,omitempty"`
+
+    	// BaseImage overrides the default ko base image.
+    	// Corresponds to, and overrides, the `defaultBaseImage` in `.ko.yaml`.
+    	BaseImage string `yaml:"fromImage,omitempty"`
+
+    	// Env are environment variables, in the `key=value` form, passed to the build.
+    	// For example: `CGO_ENABLED=1`.
+    	Env []string `yaml:"env,omitempty"`
+
+    	// Platforms is the list of platforms to build images for. Each platform
+    	// is of the format `os/arch[/variant]`, e.g., `linux/amd64`.
+    	// By default, the ko builder builds for `all` platforms supported by the
+    	// base image.
+    	Platforms []string `yaml:"platforms,omitempty"`
+
+    	// SourceDateEpoch is the `created` time of the container image.
+    	// Specify as the number of seconds since January 1st 1970, 00:00 UTC.
+    	// You can override this value by setting the `SOURCE_DATE_EPOCH`
+    	// environment variable.
+    	SourceDateEpoch unit64 `yaml:"sourceDateEpoch,omitempty"`
+    }
+    ```
+
+3.  Add a `KoArtifact` field to the `ArtifactType` struct:
+
+    ```go
+    type ArtifactType struct {
+    	// KoArtifact builds images using [ko](https://github.com/google/ko).
+    	KoArtifact *KoArtifact `yaml:"ko,omitempty" yamltags:"oneOf=artifact"`
+    }
+    ```
+
+Example basic config, this will be sufficient for many users:
+
+```yaml
+apiVersion: skaffold/v2beta14
+kind: Config
+build:
+  artifacts:
+  - image: ko://example.com/helloworld
+    ko: {}
+```
+
+The value of the `image` field is the Go import path of the app entry point,
+[prefixed by `ko://`](https://github.com/google/ko/pull/58).
+
+ko requires setting a
+[`KO_DOCKER_REPO`](https://github.com/google/ko#choose-destination)
+environment variable. This determines where container images are pushed.
+The Skaffold
+[default repo](https://skaffold.dev/docs/environment/image-registries/)
+maps directly to this value.
+
+### Open Questions
+
+1.  Should Skaffold embed ko (as a Go module), or shell out?
+
+    Benefits of embedding:
```

Comment:
one more benefit: it doesn't require users to install/know anything about `ko` to adopt it. this makes it easier for users to try it out simply by changing around some config in their `skaffold.yaml`, rather than installing a new tool entirely.

Reply:
It would be great to do it under the hood  by prompting the user without actually them changing the artifact type. We can easily do it for IDE users. For CLI users, we can show this as a tip in `skaffold diagnose`

Reply:
@tejal29 do you mean as part of `skaffold init`?